plugins/sparc2_alignment_data_collector.py

```diff
@@ -0,0 +1,699 @@
+        self.cancel_requested = False
+        self.start_btn.Disable()
+        self.pause_btn.Enable()
+        self.cancel_btn.Enable()
+        self.status_lbl.SetLabel("Status: Running...")
+        self.n_images = nl * ns * nz
+        self.progress.SetRange(self.n_images)
+        self.progress.SetValue(0)
+
+        self._aligned_pos = {
+            "l": self.mirror.position.value["l"],
+            "s": self.mirror.position.value["s"],
+            "x": self.mirror_xy.position.value["x"],
+            "y": self.mirror_xy.position.value["y"],
+            "z": self.stage.position.value["z"],
+        }
+
+        thread = threading.Thread(
+            target=self._run_acquisition,
+            args=(
+                minus_dl,
+                plus_dl,
+                minus_ds,
+                plus_ds,
+                minus_dz,
+                plus_dz,
+                nl,
+                ns,
+                nz,
+                path,
+            ),
+            daemon=True,
+        )
+        thread.start()
+
+    def _on_cancel(self, evt):
+        self.cancel_requested = True
+        self.status_lbl.SetLabel("Status: Cancel requested...")
+        self.cancel_btn.Disable()
+
+    def _restore_position(self):
+        if self._aligned_pos:
+            self.mirror.moveAbs({"l": self._aligned_pos["l"]}).result()
+            self.mirror.moveAbs({"s": self._aligned_pos["s"]}).result()
+            self.stage.moveAbs({"z": self._aligned_pos["z"]}).result()
+
+    def _run_acquisition(
+        self, minus_dl, plus_dl, minus_ds, plus_ds, minus_dz, plus_dz, nl, ns, nz, path
+    ):
+        l0 = self._aligned_pos["l"]
+        s0 = self._aligned_pos["s"]
+        z0 = self._aligned_pos["z"]
+        l_min, l_max = self.mirror.axes["l"].range
+        s_min, s_max = self.mirror.axes["s"].range
+        z_min, z_max = self.stage.axes["z"].range
+
+        l_values = numpy.clip(numpy.linspace(l0 - minus_dl, l0 + plus_dl, nl), l_min, l_max)
+        s_values = numpy.clip(numpy.linspace(s0 - minus_ds, s0 + plus_ds, ns), s_min, s_max)
+        z_values = numpy.clip(numpy.linspace(z0 - minus_dz, z0 + plus_dz, nz), z_min, z_max)
+
+        visited_l, visited_s, visited_z = [], [], []
+
+        try:
+            idx = 0
+            for l in l_values:
+                if self.cancel_requested:
+                    raise CancelledError("Cancelled by user.")
+                self.mirror.moveAbs({"l": l}).result()
+                for s in s_values:
+                    if self.cancel_requested:
+                        raise CancelledError("Cancelled by user.")
+                    self.mirror.moveAbs({"s": s}).result()
+                    for z in z_values:
+                        if self.cancel_requested:
+                            raise CancelledError("Cancelled by user.")
+                        if self._pause_requested:
+                            wx.CallAfter(self.status_lbl.SetLabel, "Status: Paused")
```

Comment:
Shouldn't it first move to the previously known aligned position? This way, if you do pause/resume, you don't automatically mess up the acquisition by having a wrong "new" aligned position.
Also, you might want to make it clearer to the user that after resuming the position has to be *aligned*. Maybe set the status to "Status: paused on aligned position"?

Reply:
Right, will change the status to "Status: paused on aligned position"

The call in line 580 `wx.CallAfter(self._restore_position)` does move it to the previously know aligned position and then waits for resume in line 584 `self._resume_requested.wait()`